man/man1/gfrcmd.1
.\" This manpage has been automatically generated by docbook2man 
.\" from a DocBook document.  This tool can be found at:
.\" <http://shell.ipoline.com/~elmert/comp/docbook2X/> 
.\" Please send any bug reports, improvements, comments, patches, 
.\" etc. to Steve Cheng <steve@ggi-project.org>.
.TH "GFRCMD" "1" "12 November 2006" "Gfarm" ""

.SH NAME
gfrcmd \- Gfarm remote shell
.SH SYNOPSIS

\fBgfrcmd\fR \fB\fIhost\fB\fR \fB\fIcommand\fB\fR\fI ...\fR

.SH "DESCRIPTION"
.PP
\fBgfrcmd\fR logs into a filesystem node \fIhost\fR and executes
\fIcommand\fRs on the filesystem node.  It provides a similar
functionality to rsh.  The following is a list of differences, however.
.PP
A remote server is not executed by inetd, but is incorporated with
gfsd(8).
.PP
The user is authenticated and authorized by the Gfarm security
mechanism.  Since this does not rely on privileged TCP ports  for
authentication like rsh, the maximum number of remote executions is
not limited by the number of privileged ports.
.PP
Unlike rsh, \fBgfrcmd\fR utilizes a single TCP connection to a
filesystem node.  It requires less kernel resources than rsh when
several processes are executed.
.PP
\fBgfrcmd\fR inherits the environment variable DISPLAY and the
authentication information of the X Window System.
.PP
On the filesystem node, the current working directory is not the home
directory that is often shared using NFS, but a spool directory of the
Gfarm filesystem, which makes it possible to leave core files on each
filesystem node when debugging.
.PP
It has functionality that invokes a debugger when a fatal signal
occurs.
.SH "OPTIONS"
.TP
\fB-l\fIusername\fB\fR
Specifies a username on a remote host.
.TP
\fB-n\fR
Redirects the standard input from the special device /dev/null.
.TP
\fB-r\fR
Executes a remote command directly without using a login shell.
.TP
\fB-y\fR
Inherits the environment variable, DISPLAY.
.TP
\fB-X\fR
Inherits the authentication information of the X Window System.  It
is useful in a situation where the home directory is not
shared.
.TP
\fB-v\fR
Displays error message verbosely.
.TP
\fB-?\fR
Displays a list of command options.
.SH "ENVIRONMENT VARIABLES"
.TP
\fBGFARM_DEBUG_MODE\fR
Specifies the debugger command that is invoked when a remotely
executed command aborts with a fatal signal.  When gdb is specified,
xterm and the GNU debugger gdb are executed.
To utilize this functionality, it is necessary to invoke \fBgfrcmd\fR
on the X Window System.
.SH "BUGS"
.PP
The -l option is not implemented yet.


ChangeLog
2006-11-13  Noriyuki Soda  <soda@sra.co.jp>

	* doc/KNOWN_PROBLEMS.ja: fix te-ni-wo-ha

	* doc/Gfarm-FAQ.en, doc/Gfarm-FAQ.ja, doc/KNOWN_PROBLEMS.en, doc/KNOWN_PROBLEMS.ja:
	describe more detailed condtion of the nss_ldap problem:
	system-supplied pthread library is the cause.

2006-11-12  Noriyuki Soda  <soda@sra.co.jp>

	* doc/Gfarm-FAQ.en, doc/Gfarm-FAQ.ja, doc/KNOWN_PROBLEMS.en, doc/KNOWN_PROBLEMS.ja:
	clarify the nss_ldap problem,
	from [gfarm-developers:02236] and [gfarm-developers:02238]

	* INSTALL.en, INSTALL.ja:
	describe conflicts between OpenLDAP and Globus on FreeBSD,
	as reported in [gfarm-developers:02236]

	* doc/README.hook.en, doc/README.hook.ja:
	add limitations found in [gfarm-developers:02231]

	* doc/README.hook.en: recover changes which were lost in revision 1.48

	* man/ja/man1/gfrun.1, man/man1/gfrun.1, doc/html/en/ref/man1/gfrun.1.html, doc/html/ja/ref/man1/gfrun.1.html:
	regen

2006-11-12  Tohru Sotoyama  <sotoyama@sra.co.jp>

	* doc/Gfarm-FAQ.en, doc/KNOWN_PROBLEMS.en, doc/README.hook.en, doc/SETUP.detail.en:
	change `file system' to `filesystem' for uniformity

2006-11-12  Noriyuki Soda  <soda@sra.co.jp>

	* doc/docbook/en/ref/man1/gfrun.1.docbook, doc/docbook/ja/ref/man1/gfrun.1.docbook:
	add the answer for question 2) which was asked in the recent Gfarm seminar.
	from [gfarm-developers:02122]

	* doc/Gfarm-FAQ.en, doc/Gfarm-FAQ.ja:
	add answers for question 1) and 6) which were asked in the recent Gfarm seminar

2006-11-12  Tohru Sotoyama  <sotoyama@sra.co.jp>

	* doc/Gfarm-FAQ.en, doc/KNOWN_PROBLEMS.en, doc/README.hook.en, doc/SETUP.detail.en:
	correct errors pointed out in native checked file "doc-RewrittenNov06.doc"
	attached to [gfarm-developers:02171]

	* INSTALL.RPM.en, OVERVIEW.en, SETUP.en:
	change `file system' to `filesystem' for uniformity

	* doc/html/en/ref/man1/gfarm_agent.1.html, doc/html/en/ref/man1/gfdf.1.html, doc/html/en/ref/man1/gfexec.1.html, doc/html/en/ref/man1/gfhost.1.html, doc/html/en/ref/man1/gfkey.1.html, doc/html/en/ref/man1/gfps.1.html, doc/html/en/ref/man1/gfrcmd.1.html, doc/html/en/ref/man1/gfrep.1.html, doc/html/en/ref/man1/gfrm.1.html, doc/html/en/ref/man1/gfront.1.html, doc/html/en/ref/man1/gfrun.1.html, doc/html/en/ref/man1/gfsched.1.html, doc/html/en/ref/man1/gfstat.1.html, doc/html/en/ref/man1/gfusage.1.html, doc/html/en/ref/man1/gfwhere.1.html, doc/html/en/ref/man1/gfwhoami.1.html, doc/html/en/ref/man3/gfs_chmod.3.html, doc/html/en/ref/man3/gfs_mkdir.3.html, doc/html/en/ref/man3/gfs_pio_create.3.html, doc/html/en/ref/man3/gfs_pio_open.3.html, doc/html/en/ref/man3/gfs_realpath.3.html, doc/html/en/ref/man3/gfs_rename.3.html, doc/html/en/ref/man3/gfs_rmdir.3.html, doc/html/en/ref/man3/gfs_unlink.3.html, doc/html/en/ref/man3/gfs_unlink_section.3.html, doc/html/en/ref/man5/gfarm.conf.5.html, doc/html/en/ref/man8/gfmd.8.html, doc/html/en/ref/man8/gfsd.8.html, man/man1/gfarm_agent.1, man/man1/gfdf.1, man/man1/gfexec.1, man/man1/gfhost.1, man/man1/gfkey.1, man/man1/gfps.1, man/man1/gfrcmd.1, man/man1/gfrep.1, man/man1/gfrm.1, man/man1/gfront.1, man/man1/gfrun.1, man/man1/gfsched.1, man/man1/gfstat.1, man/man1/gfusage.1, man/man1/gfwhere.1, man/man1/gfwhoami.1, man/man3/gfs_chmod.3, man/man3/gfs_mkdir.3, man/man3/gfs_pio_create.3, man/man3/gfs_pio_open.3, man/man3/gfs_realpath.3, man/man3/gfs_rename.3, man/man3/gfs_rmdir.3, man/man3/gfs_unlink.3, man/man3/gfs_unlink_section.3, man/man5/gfarm.conf.5, man/man8/gfmd.8, man/man8/gfsd.8:
	regen

	* doc/docbook/en/ref/man1/gfarm_agent.1.docbook, doc/docbook/en/ref/man1/gfexec.1.docbook, doc/docbook/en/ref/man1/gfrm.1.docbook, doc/docbook/en/ref/man1/gfront.1.docbook, doc/docbook/en/ref/man1/gfstat.1.docbook, doc/docbook/en/ref/man3/gfs_chmod.3.docbook, doc/docbook/en/ref/man3/gfs_chown.3.docbook, doc/docbook/en/ref/man3/gfs_mkdir.3.docbook, doc/docbook/en/ref/man3/gfs_pio_create.3.docbook, doc/docbook/en/ref/man3/gfs_pio_open.3.docbook, doc/docbook/en/ref/man3/gfs_realpath.3.docbook, doc/docbook/en/ref/man3/gfs_rename.3.docbook, doc/docbook/en/ref/man3/gfs_rmdir.3.docbook, doc/docbook/en/ref/man3/gfs_unlink.3.docbook, doc/docbook/en/ref/man3/gfs_unlink_section.3.docbook:
	change `file system' to `filesystem' for uniformity

	* doc/docbook/en/ref/man5/gfarm.conf.5.docbook:
	make corrections missed in previous 2 revisions

	* doc/docbook/en/ref/man5/gfarm.conf.5.docbook:
	make corrections missed in previous revision

2006-11-12  Noriyuki Soda  <soda@sra.co.jp>

	* package/freebsd/Makefile:
	add "--without-openldap", if WITH_GLOBUS is specified.
	Without this fix, gfarm binaries dump core, when GSI is used.

2006-11-12  Tohru Sotoyama  <sotoyama@sra.co.jp>

	* doc/docbook/en/ref/man8/gfsd.8.docbook:
	make corrections missed in previous revision

	* doc/docbook/en/ref/man1/gfrun.1.docbook:
	make a correction missed in previous revision

	* doc/docbook/en/ref/man1/gfrep.1.docbook, doc/docbook/en/ref/man1/gfrcmd.1.docbook, doc/docbook/en/ref/man8/gfmd.8.docbook, doc/docbook/en/ref/man1/gfhost.1.docbook:
	make corrections missed in previous revision

	* doc/docbook/en/ref/man1/gfsched.1.docbook, doc/docbook/en/ref/man1/gfstat.1.docbook, doc/docbook/en/ref/man1/gfusage.1.docbook, doc/docbook/en/ref/man1/gfwhere.1.docbook, doc/docbook/en/ref/man1/gfwhoami.1.docbook, doc/docbook/en/ref/man8/gfsd.8.docbook, doc/docbook/en/ref/man1/gfps.1.docbook, doc/docbook/en/ref/man1/gfrcmd.1.docbook, doc/docbook/en/ref/man1/gfrep.1.docbook, doc/docbook/en/ref/man1/gfrm.1.docbook, doc/docbook/en/ref/man1/gfrun.1.docbook, doc/docbook/en/ref/man8/gfmd.8.docbook:
	correct errors pointed out in native checked file "manRewrittenNov06.doc"
	attached to [gfarm-developers:02194]

2006-11-12  Yusuke Tanimura  <yusuke.tanimura@aist.go.jp>

	* RELNOTES: add the bugzilla number

2006-11-12  Tohru Sotoyama  <sotoyama@sra.co.jp>

	* doc/docbook/en/ref/man1/gfhost.1.docbook, doc/docbook/en/ref/man1/gfexec.1.docbook:
	update modified date

	* doc/docbook/en/ref/man1/gfkey.1.docbook:

gftool/gfrcmd/gfrcmd.c
#include <sys/types.h>
#include <sys/socket.h>
#include <netinet/in.h>
#include <netdb.h>
#include <sys/stat.h>
#include <stdio.h>
#include <stdlib.h>
#include <string.h>
#include <unistd.h>
#include <signal.h>
#include <fcntl.h>
#include <libgen.h>
#include <gfarm/gfarm.h>

#include "gfutil.h"

#include "gfs_client.h"
#include "auth.h"
#include "config.h"

char *program_name = "gfrcmd";

void
usage()
{
	fprintf(stderr, "Usage: %s [option] host command...\n", program_name);
	fprintf(stderr, "option:\n");
	fprintf(stderr, "\t-n: disable standard input.\n");
	fprintf(stderr, "\t-r: execute a remote command directly.\n");
	fprintf(stderr, "\t-y: inherits the environment variable DISPLAY.\n");
	fprintf(stderr, "\t-X: inherits the authentication info of the "
			"X Window System.\n");
#ifdef HAVE_GSI
	fprintf(stderr, "\t-N <hostname>: "
		"use this name to authenticate peer.\n");
	fprintf(stderr, "\t-v: display GSS minor status error.\n");
#endif
	exit(1);
}

char *opt_username; /* ignored for now */
int opt_no_stdin = 0;
int opt_raw_command = 0;
int opt_xdpy_env = 0;
int opt_xauth_copy = 0;
char *opt_hostname = NULL;
int opt_auth_verbose = 0;

void
parse_option(int *argcp, char ***argvp)
{
	int argc;
	char **argv, *s;

	argc = *argcp;
	argv = *argvp;
	for (; argc > 0 && argv[0][0] == '-'; argc--, argv++) {
		for (s = &argv[0][1]; *s; s++) {
			switch (*s) {
			case 'l':
				/* XXX: FIXME. `opt_username' isn't used yet */
				if (s[1]) {
					opt_username = &s[1];
					s += strlen(s) - 1;
				} else if (argc > 1) {
					opt_username = argv[1];
					argc--, argv++;
					s = argv[0] + strlen(argv[0]) - 1;
				} else {
					usage();
				}
				break;
			case 'n':
				opt_no_stdin = 1;
				break;
			case 'r':
				opt_raw_command = 1;
				break;
			case 'y':
				opt_xdpy_env = 1;
				break;
			case 'X':
				opt_xauth_copy = 1;
				break;
			case 'N':
				if (s[1]) {
					opt_hostname = &s[1];
					s += strlen(s) - 1;
				} else if (argc > 1) {
					opt_hostname = argv[1];
					argc--, argv++;
					s = argv[0] + strlen(argv[0]) - 1;
				} else {
					usage();
				}
				break;
			case 'v':
				opt_auth_verbose = 1;
				break;
			case '?':
			default:
				usage();
			}
		}
	}
	*argcp = argc;
	*argvp = argv;
}

char *
concat(int argc, char **argv)
{
	int i, len = 0;
	char *s;

	for (i = 0; i < argc; i++)
		len += strlen(argv[i]);

	len += argc + 1;
	GFARM_MALLOC_ARRAY(s, len);
	if (s == NULL) {
		fprintf(stderr, "%s: no memory\n", program_name);
		exit(1);
	}
	strcpy(s, argv[0]);
	for (i = 1; i < argc; i++) {
		strcat(s, " ");
		strcat(s, argv[i]);
	}
	return (s);
}

int passing_signals[] = { SIGINT, SIGQUIT, SIGTERM };
volatile int send_signal = 0;

void
record_signal(int sig)
{
	send_signal = sig;
}

int
main(argc, argv)
	int argc;
	char **argv;
{
	char *e, *user, *if_hostname, *command;
	char *args[2];
	char *envs[2];
	char **argp, **envp = NULL;
	struct gfs_connection *gfs_server;
	int i, remote_pid, sig, status, coredump;
	struct hostent *hp;
	struct sockaddr_in peer_addr;

	if (argc <= 0)
		usage();

	program_name = basename(argv[0]);
	argc--, argv++;

	parse_option(&argc, &argv);
	if (argc <= 0)
		usage();

	if_hostname = argv[0];
	argc--, argv++;

	parse_option(&argc, &argv);
	if (argc <= 0)
		usage();

	e = getenv("GFARM_DEBUG_MODE");
	if (e != NULL && strcmp(e, "gdb") == 0) {
		opt_xauth_copy = 1;
		envs[0] = "GFARM_DEBUG_MODE=gdb";
		envs[1] = NULL;
		envp = envs;
	}

	if (opt_raw_command) {
		argp = argv;
		command = argv[0];
	} else {
		args[0] = concat(argc, argv);
		args[1] = NULL;
		argp = args;
		command = args[0];
	}

	/*
	 * Because gflog_auth_set_verbose() is called here, the call of
	 * gfarm_gsi_client_initialize() below may display verbose
	 * error messages, which will be displayed later again in
	 * gfarm_auth_request_gsi().
	 * But unless it's called here, some error messages (e.g. 
	 * errors in gfarmAuthInitialize()) won't be displayed at all.
	 * Thus, we go the noisy way, because that's better than nothing.
	 */
	if (opt_auth_verbose)
		gflog_auth_set_verbose(1);

	/*
	 * initialization
	 *
	 * XXX
	 * The reason we don't call gfarm_initialize() here is that
	 * we'd like to avoid overhead to access meta database.
	 * but we may have to access meta database eventually
	 * for GSI DN <-> gfarm global username mapping.
	 */
	gfarm_error_initialize();
	e = gfarm_set_local_user_for_this_local_account();
	if (e == NULL)
		e = gfarm_config_read();
#ifdef HAVE_GSI /* XXX this initialization must be removed eventually */
	if (e == NULL)
		(void)gfarm_gsi_client_initialize();
#endif
	if (e == NULL)
		e = gfarm_set_global_user_for_this_local_account();
	if (e != NULL) {
		fprintf(stderr, "%s: %s\n", program_name, e);
		exit(1);
	}

	hp = gethostbyname(if_hostname);
	if (hp == NULL || hp->h_addrtype != AF_INET) {
		e = GFARM_ERR_UNKNOWN_HOST;
	} else {
		memset(&peer_addr, 0, sizeof(peer_addr));
		memcpy(&peer_addr.sin_addr, hp->h_addr,
		       sizeof(peer_addr.sin_addr));
		peer_addr.sin_family = hp->h_addrtype;
		peer_addr.sin_port = htons(gfarm_spool_server_port);
		e = gfs_client_connect(
		    opt_hostname != NULL ? opt_hostname : if_hostname, 
		    (struct sockaddr *)&peer_addr,
		    &gfs_server);
	}
	if (e != NULL) {
		fprintf(stderr, "%s: %s\n", if_hostname, e);
		exit(1);
	}

	user = gfarm_get_global_username();
	/* XXX - kluge */
	e = gfs_client_mkdir(gfs_server, user, 0755);
	e = gfs_client_chdir(gfs_server, user);

	/* for gfs_client_command_send_signal() */
	for (i = 0; i < GFARM_ARRAY_LENGTH(passing_signals); i++) {
		struct sigaction sa;

		/*
		 * DO NOT set SA_RESTART here, because we rely on the fact
		 * that select(2) breaks with EINTR.
		 * XXX - This is not so portable. Use siglongjmp() instead?
		 */
		sa.sa_handler = record_signal;
		sigemptyset(&sa.sa_mask);
		sa.sa_flags = 0;
		sigaction(passing_signals[i], &sa, NULL);
	}

	e = gfs_client_command_request(gfs_server, command, argp, envp,
	    (opt_raw_command ? 0 : GFS_CLIENT_COMMAND_FLAG_SHELL_COMMAND) |
	    (opt_no_stdin ? GFS_CLIENT_COMMAND_FLAG_STDIN_EOF : 0) |
	    (opt_xdpy_env ? GFS_CLIENT_COMMAND_FLAG_XENVCOPY : 0) |
	    (opt_xauth_copy ? GFS_CLIENT_COMMAND_FLAG_XAUTHCOPY : 0),
	    &remote_pid);
	if (e == NULL) {
		char *e2;

		while (gfs_client_command_is_running(gfs_server)) {
			e = gfs_client_command_io(gfs_server, NULL);
			if (e == NULL && send_signal != 0) {
				e = gfs_client_command_send_signal(
				    gfs_server, send_signal);
				send_signal = 0;
			}
		}
		e2 = gfs_client_command_result(gfs_server,
		    &sig, &status, &coredump);
		if (e == NULL)
			e = e2;
	}
	if (e != NULL) {
		fprintf(stderr, "%s: %s\n", argv[0], e);
		exit(1);
	}
	if (sig) {
		fprintf(stderr, "%s: signal %d received%s.\n",
		    if_hostname, sig, coredump ? " (core dumped)" : "");
		exit(255);
	}
	exit(status);
}
